libkbfs/block_retrieval_queue.go

```diff
@@ -124,13 +148,41 @@ func (brq *blockRetrievalQueue) WorkOnRequest() <-chan *blockRetrieval {
 	return ch
 }
 
-// FinalizeRequest communicates that any subsequent requestors for this block
-// won't be notified by the current worker processing it.  This must be called
-// before sending out the responses to the blockRetrievalRequests for a given
-// blockRetrieval.
-func (brq *blockRetrievalQueue) FinalizeRequest(ptr BlockPointer) {
+// FinalizeRequest s the last step of a retrieval request once a block has been
+// obtained. It removes the request from the blockRetrievalQueue, preventing
+// more requests mutating the retrieval, then calls notifyBlockRequestor for
+// all subscribed requests.
+func (brq *blockRetrievalQueue) FinalizeRequest(retrieval *blockRetrieval, block Block, err error) {
 	brq.mtx.Lock()
-	defer brq.mtx.Unlock()
+	delete(brq.ptrs, retrieval.blockPtr)
+	brq.mtx.Unlock()
+
+	if block == nil {
+		for _, r := range retrieval.requests {
+			req := r
+			go func() {
+				req.doneCh <- err
+			}()
+		}
+		return
+	}
+	source := reflect.ValueOf(block).Elem()
+	for _, r := range retrieval.requests {
+		req := r
+		go func() {
+			// Copy the decrypted block to the caller
+			dest := reflect.ValueOf(req.block).Elem()
+			dest.Set(source)
```

Comment:
Any idea what happens here if a subsequent caller passed in a different type of block from the original caller?  Not sure if that's something we should bother protecting against, but a silent failure might be bad.